power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -74,25 +59,26 @@ std::pair<std::vector<int8_t>, bool> count_flow_sensors(MeasuredValues<sym> cons
                 }
             }
         }
-        // check voltage sensor
+        // diagonal for voltage phasor sensors
         if (measured_values.has_voltage(row) && measured_values.has_angle_measurement(row)) {
             has_at_least_one_sensor = true;
+            voltage_phasor_sensors[row] = 1;
         }
         // the system could be ill-conditioned if there is no flow sensor for one bus, except the last bus
         if (!has_at_least_one_sensor && row != n_bus - 1) {
             possibly_ill_conditioned = true;
         }
     }
-    return result;
+    return std::make_tuple(std::move(flow_sensors), std::move(voltage_phasor_sensors), possibly_ill_conditioned);
 }
 
 // re-organize the flow sensor for radial grid without phasor measurement
 // this mutate the flow sensor vector to try to assign injection sensor to branch sensor
 // all the branch should be measured if the system is observable
 // this is a sufficient condition check
 // if the grid is not radial, the behavior is undefined.
-inline void assign_injection_sensor_radial(YBusStructure const& y_bus_structure, std::vector<int8_t>& flow_sensors) {
-    Idx const n_bus = std::ssize(y_bus_structure.row_indptr) - 1;
+inline void assign_independent_sensors_radial(YBusStructure const& y_bus_structure, std::vector<int8_t>& flow_sensors,
+                                              std::vector<int8_t>& voltage_phasor_sensors, Idx const n_bus) {
```

Comment:
why does `n_bus` need to be an input variable? if you need it as one, please add an assertion to verify no undefined behavior is triggered here.

```suggestion
                                              std::vector<int8_t>& voltage_phasor_sensors, Idx const n_bus) {
    assert(n_bus == std::ssize(y_bus_structure.row_indptr) - 1);
```